cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/trigger/ProgramStatusTrigger.java

```diff
@@ -45,11 +53,71 @@ public ProgramStatusTrigger(ProgramId programId, ProgramStatus... programStatuse
 
   @Override
   public boolean isSatisfied(List<Notification> notifications) {
-    return true;
+    for (Notification notification : notifications) {
+      if (!Notification.Type.PROGRAM_STATUS.equals(notification.getNotificationType())) {
+        continue;
+      }
+      String programRunIdString = notification.getProperties().get(ProgramOptionConstants.PROGRAM_RUN_ID);
+      String programRunStatusString = notification.getProperties().get(ProgramOptionConstants.PROGRAM_STATUS);
+      // Ignore notifications which specify an invalid programRunId or programStatus
+      if (programRunIdString == null || programRunStatusString == null) {
+        continue;
+      }
+      ProgramStatus programStatus;
+      try {
+        programStatus = ProgramRunStatus.toProgramStatus(ProgramRunStatus.valueOf(programRunStatusString));
+      } catch (IllegalArgumentException e) {
+        // Return silently, this happens for statuses that are not meant to be scheduled
+        continue;
+      }
+      ProgramRunId programRunId = GSON.fromJson(programRunIdString, ProgramRunId.class);
+      ProgramId triggeringProgramId = programRunId.getParent();
+      if (this.programId.equals(triggeringProgramId) && programStatuses.contains(programStatus)) {
+        return true;
+      }
+    }
+    return false;
   }
 
   @Override
   public Set<String> getTriggerKeys() {
     return Schedulers.triggerKeysForProgramStatus(programId, programStatuses);
   }
+
+  @Override
+  public TriggerInfo getTriggerInfo(TriggerInfoContext context) {
+    for (Notification notification : context.getNotifications()) {
+      if (!Notification.Type.PROGRAM_STATUS.equals(notification.getNotificationType())) {
+        continue;
+      }
+      String programRunIdString = notification.getProperties().get(ProgramOptionConstants.PROGRAM_RUN_ID);
+      String programRunStatusString = notification.getProperties().get(ProgramOptionConstants.PROGRAM_STATUS);
+      // Ignore notifications which specify an invalid programRunId or programStatus
+      if (programRunIdString == null || programRunStatusString == null) {
+        continue;
+      }
+      ProgramStatus programStatus;
+      try {
+        programStatus = ProgramRunStatus.toProgramStatus(ProgramRunStatus.valueOf(programRunStatusString));
+      } catch (IllegalArgumentException e) {
+        // Return silently, this happens for statuses that are not meant to be scheduled
+        continue;
+      }
+      ProgramRunId programRunId = GSON.fromJson(programRunIdString, ProgramRunId.class);
+      ProgramId triggeringProgramId = programRunId.getParent();
```

Comment:
Seems like up to here this method is duplicating the code in `isSatisfied`. Better refactor them to a common private method.